# New York City Citywide Mobility Survey 2019

### This script contains the following:

#### 1. Importing Libraries and Data
#### 2. Data Cleaning

## Import libraries

In [99]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy


## Import data

In [100]:
# Define path
path = r'/Users/peimei/11-2023 New York City Citywide Mobility Survey 2019'

# Join the path using 'os.path.join'
file_path = os.path.join(path, '02 Data', 'Original Data', 'Citywide_Mobility_Survey_-_Trip_Survey_2019_20231108.csv')

# Read the CSV file into a DataFrame
trips = pd.read_csv(file_path, sep=";", index_col = False,low_memory=False)

# Display the first few rows of the DataFrame
trips.head()

hh_id trip_weight   home_cms_zone  survey_mode   person_id  person_num  \
0  19000235           0  Northern Bronx            1  1900023501           1   
1  19000235           0  Northern Bronx            1  1900023501           1   
2  19000235           0  Northern Bronx            1  1900023501           1   
3  19000235           0  Northern Bronx            1  1900023501           1   
4  19000235           0  Northern Bronx            1  1900023501           1   

   day_num travel_date  travel_date_dow        trip_id  trip_num  leg_num  \
0        1  05/30/2019                4  1900023501001         1        1   
1        2  05/31/2019                5  1900023501002         2        1   
2        2  05/31/2019                5  1900023501003         3        1   
3        3  06/01/2019                6  1900023501004         4        1   
4        3  06/01/2019                6  1900023501005         5        1   

   unlinked_trip  linked_trip_num  survey_complete depart_time arrive_time  \
0              1                1                1  05/30/2019  05/30/2019   
1              0                2                0  05/31/2019  05/31/2019   
2              0                3                0  05/31/2019  05/31/2019   
3              0                4                1  06/01/2019  06/01/2019   
4              0                5                0  06/01/2019  06/01/2019   

       o_cms_zone      d_cms_zone trip_duration trip_distance o_purpose  \
0  Manhattan Core  Southern Bronx          11.1          13.4    -9,998   
1  Southern Bronx  Northern Bronx             7           3.6        10   
2  Northern Bronx  Southern Bronx          16.7           3.4    -9,998   
3  Southern Bronx  Northern Bronx           9.1           3.7    -9,998   
4  Northern Bronx  Northern Bronx           4.4             1         1   

  o_purpose_category d_purpose d_purpose_category num_non_hh_travelers  \
0             -9,998        10                  2                    0   
1                  2    -9,998             -9,998               -9,998   
2             -9,998    -9,998             -9,998               -9,998   
3             -9,998         1                  1                    0   
4                  1    -9,998             -9,998               -9,998   

  num_hh_travelers num_travelers    hh_member_1 hh_member_2 hh_member_3  \
0                1             1  1,900,023,501           0           0   
1           -9,998        -9,998  1,900,023,501           0           0   
2           -9,998        -9,998  1,900,023,501           0           0   
3                1             1  1,900,023,501           0           0   
4           -9,998        -9,998  1,900,023,501           0           0   

  hh_member_4 hh_member_5 hh_member_6 hh_member_7 hh_member_8 hh_member_9  \
0           0           0         995         995         995         995   
1           0           0         995         995         995         995   
2           0           0         995         995         995         995   
3           0           0         995         995         995         995   
4           0           0         995         995         995         995   

  hh_member_10 hh_member_11  mode_type  mode_type_detailed  sustainable_mode  \
0          995          995          6                  11                 1   
1          995          995        995                 995               995   
2          995          995        995                 995               995   
3          995          995          7                  25                 0   
4          995          995        995                 995               995   

   mode_1  mode_2  mode_3  mode_4  driver park_location  park_pay  park_cost  \
0      30     995     995     995       1           995       995        NaN   
1  -9,998     995     995     995     995           995       995        NaN   
2  -9,998     995     995     995     995           995       

## Check for Personal Identifiable Information (PII)

In [101]:
# Check all columns for PII

trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85459 entries, 0 to 85458
Data columns (total 77 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   hh_id                       85459 non-null  int64  
 1   trip_weight                 85459 non-null  object 
 2   home_cms_zone               85459 non-null  object 
 3   survey_mode                 85459 non-null  int64  
 4   person_id                   85459 non-null  int64  
 5   person_num                  85459 non-null  int64  
 6   day_num                     85459 non-null  int64  
 7   travel_date                 85459 non-null  object 
 8   travel_date_dow             85459 non-null  int64  
 9   trip_id                     85459 non-null  int64  
 10  trip_num                    85459 non-null  int64  
 11  leg_num                     85459 non-null  int64  
 12  unlinked_trip               85459 non-null  int64  
 13  linked_trip_num             854

### Result: No PII is found. The "person_id" column does not correspond to any official personal ID number.

## Check for mixed-type data

In [102]:
for col in trips.columns.tolist():
  weird = (trips[[col]].applymap(type) != trips[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (trips[weird]) > 0:
    print (col)

o_cms_zone
d_cms_zone
trip_distance
d_distance_work
d_distance_school


In [103]:
# Modify data types of the following variables respectively

trips['trip_weight'] = trips['trip_weight'].str.replace(',', '').astype('float64')
trips['o_cms_zone'] = trips['o_cms_zone'].astype('str')
trips['d_cms_zone'] = trips['d_cms_zone'].astype('str')
trips['mode_1'] = trips['mode_1'].str.replace(',', '').astype('int64')
trips['trip_distance'] = trips['trip_distance'].str.replace(',', '').astype('float64')
trips['trip_duration'] = trips['trip_duration'].str.replace(',', '').astype('float64')
trips['d_distance_home'] = trips['d_distance_home'].str.replace(',', '').astype('float64')
trips['d_distance_work'] = trips['d_distance_work'].str.replace(',', '').astype('float64')
trips['d_distance_school'] = trips['d_distance_school'].str.replace(',', '').astype('float64')

In [104]:
# Check output

for col in trips.columns.tolist():
  weird = (trips[[col]].applymap(type) != trips[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (trips[weird]) > 0:
    print (col)
else:
        print('All good.')

All good.


## Check for unclear variable names

In [105]:
# Rename variables

trips.rename(columns = {'trip_duration':'trip_duration_min',
                           'trip_distance':'trip_distance_mile',
                           'd_distance_home':'d_distance_home_meter_imputed',
                           'd_distance_work':'d_distance_work_meter_imputed',
                           'd_distance_school':'d_distance_school_meter_imputed',
                           'o_location_type':'o_location_type_imputed',
                           'd_location_type':'d_location_type_imputed',
                           'depart_time':'depart_date',
                           'arrive_time':'arrive_date'
                          }, inplace = True)

In [106]:
# Check the output

trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85459 entries, 0 to 85458
Data columns (total 77 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   hh_id                            85459 non-null  int64  
 1   trip_weight                      85459 non-null  float64
 2   home_cms_zone                    85459 non-null  object 
 3   survey_mode                      85459 non-null  int64  
 4   person_id                        85459 non-null  int64  
 5   person_num                       85459 non-null  int64  
 6   day_num                          85459 non-null  int64  
 7   travel_date                      85459 non-null  object 
 8   travel_date_dow                  85459 non-null  int64  
 9   trip_id                          85459 non-null  int64  
 10  trip_num                         85459 non-null  int64  
 11  leg_num                          85459 non-null  int64  
 12  unlinked_trip     

## Filter out records with flags (defined by data source)
high_speed_flag: Unusually high speed, given mode <br/>
long_distance_flag: Unusually long distance, given mode <br/>
invalid_purpose_flag: Non-worker / non-student has a work/school trip purpose

In [107]:
trips['high_speed_flag'].value_counts(dropna = False)

0    80777
1     4682
Name: high_speed_flag, dtype: int64

In [108]:
trips['long_distance_flag'].value_counts(dropna = False)

0    85420
1       39
Name: long_distance_flag, dtype: int64

In [109]:
trips['invalid_purpose_flag'].value_counts(dropna = False)

0    85163
1      296
Name: invalid_purpose_flag, dtype: int64

In [110]:
trips = trips[(trips['high_speed_flag']==0) & (trips['long_distance_flag']==0) & (trips['long_distance_flag']==0)]
trips.shape

(80773, 77)

## Filter out records with incomplete survey input

In [111]:
trips = trips[trips['survey_complete']==1]
trips.head()

hh_id  trip_weight   home_cms_zone  survey_mode   person_id  \
0   19000235          0.0  Northern Bronx            1  1900023501   
3   19000235          0.0  Northern Bronx            1  1900023501   
12  19000235          0.0  Northern Bronx            1  1900023501   
13  19000235          0.0  Northern Bronx            1  1900023501   
15  19000235          0.0  Northern Bronx            1  1900023501   

    person_num  day_num travel_date  travel_date_dow        trip_id  trip_num  \
0            1        1  05/30/2019                4  1900023501001         1   
3            1        3  06/01/2019                6  1900023501004         4   
12           1        4  06/02/2019                7  1900023501013        13   
13           1        4  06/02/2019                7  1900023501014        14   
15           1        5  06/03/2019                1  1900023501016        16   

    leg_num  unlinked_trip  linked_trip_num  survey_complete depart_date  \
0         1              1                1                1  05/30/2019   
3         1              0                4                1  06/01/2019   
12        1              0               13                1  06/02/2019   
13        1              0               14                1  06/02/2019   
15        1              0               16                1  06/03/2019   

   arrive_date      o_cms_zone      d_cms_zone  trip_duration_min  \
0   05/30/2019  Manhattan Core  Southern Bronx               11.1   
3   06/01/2019  Southern Bronx  Northern Bronx                9.1   
12  06/02/2019             nan  Southern Bronx               17.9   
13  06/02/2019  Southern Bronx  Northern Bronx               23.4   
15  06/03/2019  Northern Bronx  Northern Bronx               13.7   

    trip_distance_mile o_purpose o_purpose_category d_purpose  \
0                 13.4    -9,998             -9,998        10   
3                  3.7    -9,998             -9,998         1   
12                 9.8    -9,998             -9,998        50   
13                11.1        50                  7        50   
15                 2.5    -9,998             -9,998        50   

   d_purpose_category num_non_hh_travelers num_hh_travelers num_travelers  \
0                   2                    0                1             1   
3                   1                    0                1             1   
12                  7                    1                2             3   
13                  7                    1                2             3   
15                  7                    0                1             1   

      hh_member_1    hh_member_2 hh_member_3 hh_member_4 hh_member_5  \
0   1,900,023,501              0           0           0           0   
3   1,900,023,501              0           0           0           0   
12  1,900,023,501  1,900,023,502           0           0           0   
13  1,900,023,501  1,900,023,502           0           0           0   
15  1,900,023,501              0           0           0           0   

   hh_member_6 hh_member_7 hh_member_8 hh_member_9 hh_member_10 hh_member_11  \
0          995         995         995         995          995          995   
3          995         995         995         995          995          995   
12         995         995         995         995          995          995   
13         995         995         995         995          995          995   
15         995         995         995         995          995          995   

    mode_type  mode_type_detailed  sustainable_mode  mode_1  mode_2  mode_3  \
0           6                  11                 1      30     995     995   
3           7                  25                 0       6     995     995   
12          7                  25                 0       6     995     995   
13          7                  25                 0       6     995     995   
15          2                   3                 0      18     

In [112]:
# Check output
trips['survey_complete'].value_counts(dropna = False)

1    74977
Name: survey_complete, dtype: int64

## Create a subset containing only necessary columns

In [113]:
# Drop unnecessary columns and save as a new dataframe

trips_trimmed = trips.drop(columns = ['survey_complete','person_num','unlinked_trip','linked_trip_num', 'o_purpose','o_purpose_category', 'd_purpose', 'd_purpose_category','num_non_hh_travelers', 'num_hh_travelers', 'num_travelers','hh_member_1','hh_member_2','hh_member_3','hh_member_4','hh_member_5','hh_member_6','hh_member_7', 'hh_member_8', 'hh_member_9', 'hh_member_10', 'hh_member_11', 'analyst_split', 'driver', 'park_location', 'park_pay', 'park_cost','taxi_cost','taxi_cost_amount', 'bike_park_location', 'scooter_park_location','analyst_merge', 'analyst_split_loop', 'user_merged', 'user_split', 'added_trip', 'd_distance_home_meter_imputed','d_distance_work_meter_imputed', 'd_distance_school_meter_imputed'])

In [114]:
trips_trimmed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74977 entries, 0 to 85458
Data columns (total 38 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   hh_id                       74977 non-null  int64  
 1   trip_weight                 74977 non-null  float64
 2   home_cms_zone               74977 non-null  object 
 3   survey_mode                 74977 non-null  int64  
 4   person_id                   74977 non-null  int64  
 5   day_num                     74977 non-null  int64  
 6   travel_date                 74977 non-null  object 
 7   travel_date_dow             74977 non-null  int64  
 8   trip_id                     74977 non-null  int64  
 9   trip_num                    74977 non-null  int64  
 10  leg_num                     74977 non-null  int64  
 11  depart_date                 74977 non-null  object 
 12  arrive_date                 74977 non-null  object 
 13  o_cms_zone                  749

In [115]:
# Check the output

trips_trimmed.shape

(74977, 38)

### Result: Number of columns is reduced from 77 to 38.

## Check for missing values

In [116]:
# Check for NaN values

trips_trimmed.isnull().sum()

hh_id                          0
trip_weight                    0
home_cms_zone                  0
survey_mode                    0
person_id                      0
day_num                        0
travel_date                    0
travel_date_dow                0
trip_id                        0
trip_num                       0
leg_num                        0
depart_date                    0
arrive_date                    0
o_cms_zone                     0
d_cms_zone                     0
trip_duration_min              0
trip_distance_mile            87
mode_type                      0
mode_type_detailed             0
sustainable_mode               0
mode_1                         0
mode_2                         0
mode_3                         0
mode_4                         0
used_transit                   0
transit_access                 0
transit_egress                 0
high_speed_flag                0
long_distance_flag             0
invalid_purpose_flag           0
o_purpose_

### Result: The variable 'trip_distance_mile' contains 87 NaN values. Given that this accounts for a negligible portion of the 79,059 records, NaN values can be imputed with zero.

## Impute NaN values with the mean value

In [117]:
mean_value = trips_trimmed['trip_distance_mile'].mean()
trips_trimmed['trip_distance_mile'].fillna(mean_value, inplace = True)

In [118]:
# Check output
trips_trimmed.isnull().sum()

hh_id                         0
trip_weight                   0
home_cms_zone                 0
survey_mode                   0
person_id                     0
day_num                       0
travel_date                   0
travel_date_dow               0
trip_id                       0
trip_num                      0
leg_num                       0
depart_date                   0
arrive_date                   0
o_cms_zone                    0
d_cms_zone                    0
trip_duration_min             0
trip_distance_mile            0
mode_type                     0
mode_type_detailed            0
sustainable_mode              0
mode_1                        0
mode_2                        0
mode_3                        0
mode_4                        0
used_transit                  0
transit_access                0
transit_egress                0
high_speed_flag               0
long_distance_flag            0
invalid_purpose_flag          0
o_purpose_imputed             0
d_purpos

In [119]:
trips_trimmed['trip_distance_mile'] = trips_trimmed['trip_distance_mile'].astype('float64')

In [120]:
trips_trimmed['trip_distance_mile'].dtype

dtype('float64')

## Check for abnormal values

In [121]:
# Set display options to show all columns without truncation
pd.set_option('display.max_columns', None)

# Check descriptive statistics
trips_trimmed.describe()

hh_id   trip_weight   survey_mode     person_id       day_num  \
count  7.497700e+04  74977.000000  74977.000000  7.497700e+04  74977.000000   
mean   2.587448e+07    349.735480      1.030663  2.587448e+09      3.818051   
std    3.285161e+07   1649.406097      0.192368  3.285161e+09      2.043527   
min    1.900024e+07      0.000000      1.000000  1.900024e+09      1.000000   
25%    1.925115e+07     15.600000      1.000000  1.925115e+09      2.000000   
50%    1.951710e+07     33.000000      1.000000  1.951710e+09      4.000000   
75%    1.979035e+07    118.200000      1.000000  1.979035e+09      6.000000   
max    1.999892e+08  94812.000000      3.000000  1.999892e+10      7.000000   

       travel_date_dow       trip_id      trip_num       leg_num  \
count     74977.000000  7.497700e+04  74977.000000  74977.000000   
mean          3.976473  2.587448e+12     22.286181     28.078264   
std           1.955956  3.285161e+12     17.461812    161.139255   
min           1.000000  1.900024e+12      1.000000      1.000000   
25%           2.000000  1.925115e+12      9.000000      1.000000   
50%           4.000000  1.951710e+12     19.000000      1.000000   
75%           6.000000  1.979035e+12     32.000000      1.000000   
max           7.000000  1.999892e+13    178.000000    995.000000   

       trip_duration_min  trip_distance_mile     mode_type  \
count       74977.000000        74977.000000  74977.000000   
mean           24.086732            3.635827      7.272337   
std           102.100557           10.676397      4.038649   
min             0.100000            0.000000      1.000000   
25%             5.600000            0.300000      6.000000   
50%            11.100000            1.000000      7.000000   
75%            22.600000            3.600000      9.000000   
max         17282.000000         1001.200000    995.000000   

       mode_type_detailed  sustainable_mode        mode_1        mode_2  \
count        74977.000000      74977.000000  74977.000000  74977.000000   
mean            30.890726          0.660669     26.551409    949.730344   
std             15.917597          3.662714    121.109960    204.113513   
min              1.000000          0.000000  -9998.000000      1.000000   
25%             16.000000          0.000000      1.000000    995.000000   
50%             26.000000          1.000000      6.000000    995.000000   
75%             48.000000          1.000000     30.000000    995.000000   
max            995.000000        995.000000    997.000000    997.000000   

             mode_3        mode_4  used_transit  high_speed_flag  \
count  74977.000000  74977.000000  74977.000000          74977.0   
mean     989.168705    994.006909      0.250490              0.0   
std       74.576605     30.790926      0.433298              0.0   
min        1.000000      1.000000      0.000000              0.0   
25%      995.000000    995.000000      0.000000              0.0   
50%      995.000000    995.000000      0.000000              0.0   
75%      995.000000    995.000000      1.000000              0.0   
max      997.000000    995.000000      1.000000              0.0   

       long_distance_flag  invalid_purpose_flag  o_location_type_imputed  \
count             74977.0          74977.000000             74977.000000   
mean                  0.0              0.003761                 3.166571   
std                   0.0              0.061213                 1.282582   
min                   0.0              0.000000                 1.000000   
25%                   0.0              0.000000                 2.000000   
50%                   0.0              0.000000                 4.000000   
75%                   0.0              0.000000                 4.000000   
max                   0.0              1.000000                 4.000000   

       d_location_type_imputed  o_congestion  d_congestion  
count             74977.000000  74977.000000  74977.000000  
mean                  3.175641    

### Results 
* Range: <br/>
&nbsp;1.) 'trip_duration_min' has an abnormally wide range of 0 - 17282. To be investigated!<br/>
&nbsp;2.) 'trip_distance_mile' has an abnormally wide range of 0 - 7386.8. To be investigated!<br/>
&nbsp;3.) These columns have an abnormally wide range of 1 - 995: 'mode_type', 'leg_num', 'mode_type', 'mode_type_detailed', 'sustainable_mode' 
&nbsp;According to the Survey User Guide, 995 is code for "a value or response not required under the circumstances." <br/>
&nbsp;4.) These columns have an abnormally wide range of 1 - 997: 'mode_2', 'mode_3', 'mode_4'<br/>
&nbsp;According to the Data Dictionary, 997 is code for "Other." <br/><br/>

## Remove outliers

In [122]:
# Check how many trips last zero minutes
trips_duration_zero = trips_trimmed[trips_trimmed['trip_duration_min'] == 0]
trips_duration_zero.shape[0]

0

In [123]:
# Check how many trips last more than 24 hours
trips_more_than_24hr = trips_trimmed[trips_trimmed['trip_duration_min'] > 24*60]
trips_more_than_24hr.shape[0]

53

In [124]:
# Check how many trips are zero miles long
trips_distance_zero = trips_trimmed[trips_trimmed['trip_distance_mile'] == 0]
trips_distance_zero.shape[0]

158

In [125]:
# Assume the longest travel distance by any transportation is 1000 miles a day.
# Check how many trips are longer than 1000 miles
trips_more_than_1000miles = trips_trimmed[trips_trimmed['trip_distance_mile'] > 1000]

In [126]:
trips_more_than_1000miles.shape[0]

1

### As the sample size won't be significantly reduced by applying these four conditions, I decided to <br/>1) remove outliers by limiting the trip duration to 1440 minutes & the trip distance to 1000 miles and <br/>2) remove zeros in both columns.

In [127]:
trips_trimmed_2 = trips_trimmed[(trips_trimmed['trip_duration_min'] <= 24*60) & 
    (trips_trimmed['trip_duration_min'] != 0) & (trips_trimmed['trip_distance_mile'] <= 1000)]

In [128]:
trips_trimmed_2.shape

(74923, 38)

In [129]:
# Check output
trips_trimmed_2.describe()

hh_id   trip_weight   survey_mode     person_id       day_num  \
count  7.492300e+04  74923.000000  74923.000000  7.492300e+04  74923.000000   
mean   2.587197e+07    349.336639      1.030685  2.587197e+09      3.818320   
std    3.284534e+07   1648.236984      0.192435  3.284534e+09      2.043738   
min    1.900024e+07      0.000000      1.000000  1.900024e+09      1.000000   
25%    1.925115e+07     15.600000      1.000000  1.925115e+09      2.000000   
50%    1.951710e+07     33.000000      1.000000  1.951710e+09      4.000000   
75%    1.979035e+07    118.200000      1.000000  1.979035e+09      6.000000   
max    1.999892e+08  94812.000000      3.000000  1.999892e+10      7.000000   

       travel_date_dow       trip_id      trip_num       leg_num  \
count     74923.000000  7.492300e+04  74923.000000  74923.000000   
mean          3.976616  2.587197e+12     22.291886     28.097714   
std           1.955830  3.284534e+12     17.464401    161.195686   
min           1.000000  1.900024e+12      1.000000      1.000000   
25%           2.000000  1.925115e+12      9.000000      1.000000   
50%           4.000000  1.951710e+12     19.000000      1.000000   
75%           6.000000  1.979035e+12     32.000000      1.000000   
max           7.000000  1.999892e+13    178.000000    995.000000   

       trip_duration_min  trip_distance_mile     mode_type  \
count       74923.000000        74923.000000  74923.000000   
mean           22.430295            3.612641      7.272813   
std            55.273827            9.992449      4.039783   
min             0.100000            0.000000      1.000000   
25%             5.600000            0.300000      6.000000   
50%            11.100000            1.000000      7.000000   
75%            22.500000            3.600000      9.000000   
max          1440.000000          760.100000    995.000000   

       mode_type_detailed  sustainable_mode        mode_1        mode_2  \
count        74923.000000      74923.000000  74923.000000  74923.000000   
mean            30.898029          0.660652     26.528062    949.774942   
std             15.916825          3.664013    121.048922    204.018118   
min              1.000000          0.000000  -9998.000000      1.000000   
25%             16.000000          0.000000      1.000000    995.000000   
50%             26.000000          1.000000      6.000000    995.000000   
75%             48.000000          1.000000     30.000000    995.000000   
max            995.000000        995.000000    997.000000    997.000000   

             mode_3        mode_4  used_transit  high_speed_flag  \
count  74923.000000  74923.000000  74923.000000          74923.0   
mean     989.164502    994.006193      0.250284              0.0   
std       74.603312     30.802008      0.433179              0.0   
min        1.000000      1.000000      0.000000              0.0   
25%      995.000000    995.000000      0.000000              0.0   
50%      995.000000    995.000000      0.000000              0.0   
75%      995.000000    995.000000      1.000000              0.0   
max      997.000000    995.000000      1.000000              0.0   

       long_distance_flag  invalid_purpose_flag  o_location_type_imputed  \
count             74923.0          74923.000000             74923.000000   
mean                  0.0              0.003764                 3.166571   
std                   0.0              0.061235                 1.282568   
min                   0.0              0.000000                 1.000000   
25%                   0.0              0.000000                 2.000000   
50%                   0.0              0.000000                 4.000000   
75%                   0.0              0.000000                 4.000000   
max                   0.0              1.000000                 4.000000   

       d_location_type_imputed  o_congestion  d_congestion  
count             74923.000000  74923.000000  74923.000000  
mean                  3.175861    

## Check for duplicates

In [130]:
# Check for duplicates

trips_trimmed_2_dups = trips_trimmed_2[trips_trimmed_2.duplicated()]

### Result: No duplicates are found.

## Impute numeric-value codes with actual texts in some columns

In [131]:
# Create a copy of the DataFrame
trips_trimmed_2 = trips_trimmed_2.copy()

# Modify data type
trips_trimmed_2['travel_date'] = trips_trimmed_2['travel_date'].astype('datetime64[ns]')
trips_trimmed_2['depart_date'] = trips_trimmed_2['depart_date'].astype('datetime64[ns]')
trips_trimmed_2['arrive_date'] = trips_trimmed_2['arrive_date'].astype('datetime64[ns]')

In [132]:
# Imputation for column 'travel_date_dow'

trips_trimmed_2['travel_date_dow'] = trips_trimmed_2['travel_date'].dt.day_name()

In [133]:
trips_trimmed_2.head()

hh_id  trip_weight   home_cms_zone  survey_mode   person_id  day_num  \
0   19000235          0.0  Northern Bronx            1  1900023501        1   
3   19000235          0.0  Northern Bronx            1  1900023501        3   
12  19000235          0.0  Northern Bronx            1  1900023501        4   
13  19000235          0.0  Northern Bronx            1  1900023501        4   
15  19000235          0.0  Northern Bronx            1  1900023501        5   

   travel_date travel_date_dow        trip_id  trip_num  leg_num depart_date  \
0   2019-05-30        Thursday  1900023501001         1        1  2019-05-30   
3   2019-06-01        Saturday  1900023501004         4        1  2019-06-01   
12  2019-06-02          Sunday  1900023501013        13        1  2019-06-02   
13  2019-06-02          Sunday  1900023501014        14        1  2019-06-02   
15  2019-06-03          Monday  1900023501016        16        1  2019-06-03   

   arrive_date      o_cms_zone      d_cms_zone  trip_duration_min  \
0   2019-05-30  Manhattan Core  Southern Bronx               11.1   
3   2019-06-01  Southern Bronx  Northern Bronx                9.1   
12  2019-06-02             nan  Southern Bronx               17.9   
13  2019-06-02  Southern Bronx  Northern Bronx               23.4   
15  2019-06-03  Northern Bronx  Northern Bronx               13.7   

    trip_distance_mile  mode_type  mode_type_detailed  sustainable_mode  \
0                 13.4          6                  11                 1   
3                  3.7          7                  25                 0   
12                 9.8          7                  25                 0   
13                11.1          7                  25                 0   
15                 2.5          2                   3                 0   

    mode_1  mode_2  mode_3  mode_4  used_transit transit_access  \
0       30     995     995     995             1              1   
3        6     995     995     995             0            995   
12       6     995     995     995             0            995   
13       6     995     995     995             0            995   
15      18      49      30     995             1              1   

   transit_egress  high_speed_flag  long_distance_flag  invalid_purpose_flag  \
0               1                0                   0                     0   
3             995                0                   0                     0   
12            995                0                   0                     0   
13            995                0                   0                     0   
15              1                0                   0                     0   

   o_purpose_imputed d_purpose_imputed o_purpose_category_imputed  \
0             -9,998                10                     -9,998   
3                 10                 1                          2   
12            -9,998                50                     -9,998   
13                50                50                          7   
15            -9,998                50                     -9,998   

   d_purpose_category_imputed  o_location_type_imputed  \
0                           2                        4   
3                           1                        2   
12                          7                        4   
13                          7                        4   
15                          7                        4   

    d_location_type_imputed  o_congestion  d_congestion  
0                         2             1             0  
3                         1             0             0  
12                        4             0             0  
13                        4             0             0  
15                        4             0             0

In [134]:
# Imputation for column 'survey_mode'

# Define a mapping of numerical values to corresponding texts
survey_mode_mapping = {
    -9998: 'missing: non-response',
    1: 'smartphone app',
    2: 'online, browser',
    3: 'call center',
    995: 'missing, skip logic',
}

# Replace numerical values with corresponding texts
trips_trimmed_2['survey_mode'] = trips_trimmed_2['survey_mode'].replace(survey_mode_mapping)

In [135]:
trips_trimmed_2['survey_mode'].value_counts(dropna = False)

smartphone app     72897
online, browser     1753
call center          273
Name: survey_mode, dtype: int64

In [136]:
# Imputation for column 'mode_type'

# Define a mapping of numerical values to corresponding texts
mode_type_mapping = {
    -9998: 'Missing: Non-response',
    1: 'For-Hire Vehicle',
    2: 'App-Based For-Hire Vehicle',
    3: 'Commuter Rail',
    4: 'Bus',
    5: 'Ferry',
    6: 'Subway',
    7: 'Vehicle',
    8: 'Bike',
    9: 'Walk',
    10: 'Other',
    995: 'missing, skip logic',
}

# Replace numerical values with corresponding texts
trips_trimmed_2['mode_type'] = trips_trimmed_2['mode_type'].replace(mode_type_mapping)


In [137]:
trips_trimmed_2['mode_type'].value_counts(dropna = False)

Walk                          28141
Vehicle                       23314
Subway                        12661
Bus                            5235
Bike                           1327
Other                          1273
App-Based For-Hire Vehicle     1220
For-Hire Vehicle                684
Commuter Rail                   636
Ferry                           431
missing, skip logic               1
Name: mode_type, dtype: int64

In [138]:
# Imputation for column 'mode_type_detailed'
# Define a mapping of numerical values to corresponding texts for 'mode_type_detailed'
mode_type_detailed_mapping = {
    -9998: 'Missing: Non-response',
    1: 'Yellow taxi',
    2: 'Green taxi',
    3: 'Uber, Lyft, or other smartphone-app ride service',
    4: 'Lyft Shared, UberPool, Via, or other shared-ride',
    5: 'Car service',
    6: 'Commuter rail',
    7: 'Express bus',
    8: 'Staten Island ferry',
    9: 'NYC ferry',
    10: 'Other ferry',
    11: 'Subway',
    12: 'PATH train',
    13: 'Intercity rail (e.g., Amtrak)',
    14: 'Other rail',
    15: 'Select bus service',
    16: 'Local bus service',
    17: 'Paratransit/access-a-ride',
    18: 'University shuttle/bus',
    19: 'Employer shuttle/bus',
    20: 'Other private shuttle/bus (e.g., a hotel\'s, an airport\'s)',
    21: 'Intercity bus (e.g., Greyhound, Megabus)',
    22: 'Other bus',
    23: 'Commuter van/dollar van',
    24: 'Carpool',
    25: 'Household vehicle 1',
    26: 'Household vehicle 2',
    27: 'Household vehicle 3',
    28: 'Household vehicle 4',
    29: 'Household vehicle 5',
    30: 'Other vehicle in household',
    31: 'Rental car',
    32: 'Car from work',
    33: 'Friend/colleague\'s car',
    34: 'Other vehicle',
    35: 'Carshare',
    36: 'Motorcycle',
    37: 'Moped share (e.g., Revel)',
    38: 'Personal scooter or moped',
    39: 'Other scooter or moped',
    40: 'Electric bike',
    41: 'Citi Bike (standard bicycle)',
    42: 'Citi Bike (electric bicycle)',
    43: 'Dockless bike share (standard bicycle)',
    44: 'Dockless bike share (electric bicycle)',
    45: 'Other rented bicycle',
    46: 'Borrowed bicycle (e.g., from friend)',
    47: 'Standard bicycle owned by my household',
    48: 'Walk',
    49: 'Other',
    995: 'Missing: Skip logic'
}

# Assuming 'mode_type_detailed' column contains integers
trips_trimmed_2['mode_type_detailed'] = trips_trimmed_2['mode_type_detailed'].replace(mode_type_detailed_mapping)


In [139]:
trips_trimmed_2['mode_type_detailed'].value_counts(dropna = False)

Walk                                                         28141
Household vehicle 1                                          15517
Subway                                                       12610
Local bus service                                             3696
Household vehicle 2                                           2825
Friend/colleague's car                                        1560
Car from work                                                 1326
Other vehicle in household                                    1112
Uber, Lyft, or other smartphone-app ride service              1045
Other                                                          968
Standard bicycle owned by my household                         947
Express bus                                                    830
Commuter rail                                                  636
Select bus service                                             420
Yellow taxi                                                   

In [140]:
# Imputation for column 'o_purpose_category_imputed'

# Define a mapping of numeric values to corresponding texts
o_purpose_category_mapping = {
    '-9,998': 'Missing: Non-response',
    '1': 'Home',
    '2': 'Work',
    '3': 'Work-related',
    '4': 'School',
    '5': 'Escort',
    '6': 'Shop',
    '7': 'Meal',
    '8': 'Social/Recreation',
    '9': 'Errand/Other',
    '10': 'Change mode',
    '11': 'Other (Unknown)',
    '995': 'Missing: Skip logic'
}

# Replace numeric values with actual texts
trips_trimmed_2['o_purpose_category_imputed'] = trips_trimmed_2['o_purpose_category_imputed'].replace(o_purpose_category_mapping)

# Display the updated DataFrame
trips_trimmed_2['o_purpose_category_imputed']

0        Missing: Non-response
3                         Work
12       Missing: Non-response
13                        Meal
15       Missing: Non-response
                 ...          
85454        Social/Recreation
85455        Social/Recreation
85456        Social/Recreation
85457        Social/Recreation
85458        Social/Recreation
Name: o_purpose_category_imputed, Length: 74923, dtype: object

In [141]:
trips_trimmed_2['o_purpose_category_imputed'].value_counts(dropna = False)

Home                     14688
Shop                     11736
Change mode              10732
Social/Recreation         8866
Work                      6564
Meal                      5576
Errand/Other              4134
Work-related              3979
Other (Unknown)           3658
Escort                    2936
Missing: Non-response     1536
School                     518
Name: o_purpose_category_imputed, dtype: int64

In [142]:
# Imputation for column 'd_purpose_category_imputed'

# Define a mapping of numeric values to corresponding texts
d_purpose_category_mapping = {
    '-9,998': 'Missing: Non-response',
    '1': 'Home',
    '2': 'Work',
    '3': 'Work-related',
    '4': 'School',
    '5': 'Escort',
    '6': 'Shop',
    '7': 'Meal',
    '8': 'Social/Recreation',
    '9': 'Errand/Other',
    '10': 'Change mode',
    '11': 'Other (Unknown)',
    '995': 'Missing: Skip logic'
}

# Replace numeric values with actual texts
trips_trimmed_2['d_purpose_category_imputed'] = trips_trimmed_2['d_purpose_category_imputed'].replace(d_purpose_category_mapping)

# Display the updated DataFrame
trips_trimmed_2['d_purpose_category_imputed']

0                     Work
3                     Home
12                    Meal
13                    Meal
15                    Meal
               ...        
85454    Social/Recreation
85455    Social/Recreation
85456    Social/Recreation
85457    Social/Recreation
85458    Social/Recreation
Name: d_purpose_category_imputed, Length: 74923, dtype: object

In [143]:
trips_trimmed_2['d_purpose_category_imputed'].value_counts(dropna = False)

Home                 13920
Shop                 12132
Change mode          10829
Social/Recreation     9204
Work                  7039
Meal                  5749
Errand/Other          4198
Work-related          4172
Other (Unknown)       4059
Escort                3065
School                 556
Name: d_purpose_category_imputed, dtype: int64

In [144]:
# Save dataframe as csv file

trips_trimmed_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'Citywide_Mobility_Survey_Trips_Clean.csv'))

In [145]:
trips_trimmed_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74923 entries, 0 to 85458
Data columns (total 38 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   hh_id                       74923 non-null  int64         
 1   trip_weight                 74923 non-null  float64       
 2   home_cms_zone               74923 non-null  object        
 3   survey_mode                 74923 non-null  object        
 4   person_id                   74923 non-null  int64         
 5   day_num                     74923 non-null  int64         
 6   travel_date                 74923 non-null  datetime64[ns]
 7   travel_date_dow             74923 non-null  object        
 8   trip_id                     74923 non-null  int64         
 9   trip_num                    74923 non-null  int64         
 10  leg_num                     74923 non-null  int64         
 11  depart_date                 74923 non-null  datetime64